src/Haskell/Ide/Engine/Plugin/Package.hs

```diff
@@ -0,0 +1,221 @@
+
+  supportsDocChanges <- clientSupportsDocumentChanges
+
+  case Y.decodeThrow contents :: Maybe Object of
+    Just obj -> do
+        let compsMapped = mapComponentTypes (ensureObject $ mapComponents (ensureObject $ mapCompDependencies addDep)) obj
+
+        let addDepToMainLib = fromMaybe True $ do
+              Object lib <- HM.lookup "library" compsMapped
+              sourceDirs <- HM.lookup "source-dirs" lib
+              return $ isInSourceDir sourceDirs
+
+        let newPkg = if addDepToMainLib
+            then mapMainDependencies addDep compsMapped
+            else compsMapped
+
+            newPkgText = T.decodeUtf8 $ Y.encode newPkg
+
+        let numOldLines = length $ T.lines $ T.decodeUtf8 contents
+            range = J.Range (J.Position 0 0) (J.Position numOldLines 0)
+            textEdit = J.TextEdit range newPkgText
+            docUri = filePathToUri fp
+            docId = J.VersionedTextDocumentIdentifier docUri 0
+            textDocEdit = J.TextDocumentEdit docId (J.List [textEdit])
+            wsEdit =
+              if supportsDocChanges
+                then J.WorkspaceEdit Nothing (Just (J.List [textDocEdit]))
+                else J.WorkspaceEdit (Just (HM.singleton docUri (J.List [textEdit]))) Nothing
+
+        return $ IdeResultOk wsEdit
+    Nothing -> return $ IdeResultFail (IdeError PluginError "Couldn't parse package.yaml" Null)
+
+  where
+
+    mapMainDependencies :: (Value -> Value) -> Object -> Object
+    mapMainDependencies f o =
+      let g "dependencies" = f
+          g _              = id
+      in HM.mapWithKey g o
+
+    mapComponentTypes :: (Value -> Value) -> Object -> Object
+    mapComponentTypes f o =
+      let g "executables" = f
+          g "executable"  = f
+          g "benchmarks"  = f
+          g "tests"       = f
+          g _             = id
+      in HM.mapWithKey g o
+
+    mapComponents :: (Value -> Value) -> Object -> Object
+    mapComponents = HM.map
+
+    mapCompDependencies :: (Value -> Value) -> Object -> Object
+    mapCompDependencies f o =
+      let g "dependencies" = f
+          g _              = id
+          shouldMap = ("source-dirs" `notElem` HM.keys o) || hasModule
+          hasModule = maybe False isInSourceDir $ HM.lookup "source-dirs" o
+      in if shouldMap
+          then HM.mapWithKey g o
+          else o
+
+    isInSourceDir :: Value -> Bool
+    isInSourceDir (Array sourceDirs) = any containsPrefix sourceDirs
+    isInSourceDir x = containsPrefix x
+
+    containsPrefix (String s) = s `T.isPrefixOf` T.pack modulePath
+    containsPrefix _ = False
+
+    ensureObject :: (Object -> Object) -> (Value -> Value)
+    ensureObject f v = case v of
+      Object o -> Object (f o)
+      _ -> error "Not an object!"
+
+    addDep (Array deps) = Array $ fromList (String pkgName:GHC.Exts.toList deps)
+    addDep _            = error "Not an array in addDep"
+
+
+-- | Takes a cabal file and a path to a module in the dependency you want to edit.
```

Comment:
And it will lose comments and formatting?

Reply:
With cabal files it *shouldn't*. The printed cabal module after editing doesn't contain any formatting, but it tries to only apply the additions in the diff.
For package.yaml I couldn't get makeDiffResult to work with it, so it loses formatting. Interestingly enough there's another GSOC project for format preserving yaml editing: https://github.com/wisn/format-preserving-yaml/